config/management/src/validator_config.rs

```diff
@@ -1,202 +1,127 @@
-pub fn encode_validator_address(
-    address: NetworkAddress,
-    owner_account: &AccountAddress,
-    sequence_number: u64,
-    key: &Key,
-    version: KeyVersion,
-    addr_idx: u32,
-) -> Result<RawEncNetworkAddress, Error> {
-    let raw_address = encode_address(address)?;
-    let enc_address =
-        raw_address.encrypt(key, version, owner_account, sequence_number + 1, addr_idx);
-    RawEncNetworkAddress::try_from(&enc_address).map_err(|e| {
-        Error::UnexpectedError(format!(
-            "error serializing encrypted address: '{:?}', error: {}",
-            enc_address, e
-        ))
-    })
-}
+impl ValidatorConfig {
+    pub fn build_transaction(
+        &self,
+        sequence_number: u64,
+        fullnode_address: NetworkAddress,
+        validator_address: NetworkAddress,
+        reconfigure: bool,
+    ) -> Result<Transaction, Error> {
+        let mut storage = StorageWrapper::new(
+            self.validator_backend.name(),
+            &self.validator_backend.validator_backend,
+        )?;
+        let owner_account = storage.account_address(OWNER_ACCOUNT)?;
 
-// Decodes adn decrypts the validator address
-pub fn decode_validator_address(
-    address: RawEncNetworkAddress,
-    account: &AccountAddress,
-    key: &Key,
-    addr_idx: u32,
-) -> Result<NetworkAddress, Error> {
-    let enc_addr = EncNetworkAddress::try_from(&address).map_err(|e| {
-        Error::UnexpectedError(format!(
-            "Failed to decode network address {}",
-            e.to_string()
-        ))
-    })?;
-    let raw_addr = enc_addr.decrypt(key, account, addr_idx).map_err(|e| {
-        Error::UnexpectedError(format!(
-            "Failed to decrypt network address {}",
-            e.to_string()
-        ))
-    })?;
-    NetworkAddress::try_from(&raw_addr).map_err(|e| {
-        Error::UnexpectedError(format!(
-            "Failed to decode network address {}",
-            e.to_string()
-        ))
-    })
+        let consensus_key = storage.ed25519_public_from_private(CONSENSUS_KEY)?;
+        let fullnode_network_key = storage.x25519_public_from_private(FULLNODE_NETWORK_KEY)?;
+        let validator_network_key = storage.x25519_public_from_private(VALIDATOR_NETWORK_KEY)?;
+
+        // Build Validator address including protocols and encryption
+        // Append ln-noise-ik and ln-handshake protocols to base network addresses
+        // and encrypt the validator address.
+        let validator_address =
+            validator_address.append_prod_protos(validator_network_key, HANDSHAKE_VERSION);
+        let raw_validator_address = encode_address(validator_address)?;
+        // Only supports one address for now
+        let addr_idx = 0u32;
+        let key = TEST_SHARED_VAL_NETADDR_KEY;
+        let version = TEST_SHARED_VAL_NETADDR_KEY_VERSION;
+        let enc_validator_address = raw_validator_address.encrypt(
+            &key,
+            version,
+            &owner_account,
+            sequence_number + 1,
```

Comment:
ngl, this took me a while to figure out.  If only we could make this function later (current_sequence_number) instead of the sequence number we expect it to be.  But, that might cause a whole other layer of issues

Reply:
ahh, yeah, we should add a comment on why we select this magic number... :D